R/format_data.R

```diff
@@ -996,6 +996,310 @@ fmt_percent <- function(data,
+  # with this formatter
+  if (!column_classes_are_valid(
+    data = data,
+    columns = {{ columns }},
+    valid_classes = c("numeric", "integer")
+  )
+  ) {
+    stop(
+      "The `fmt_fraction()` function can only be used on `columns` with numeric data",
+      call. = FALSE
+    )
+  }
+
+  # In the inline layout (where there is no special typesetting of the
+  # fraction part) we must include a space to distinguish between the
+  # whole number and fraction parts
+  # if (layout == "inline") {
+  #   incl_space <- TRUE
+  # }
+
+  # Use locale-based marks if a locale ID is provided
+  sep_mark <- get_locale_sep_mark(locale, sep_mark, use_seps)
+  dec_mark <- get_locale_dec_mark(locale, dec_mark)
+
+  # Pass `data`, `columns`, `rows`, and the formatting
+  # functions as a function list to `fmt()`
+  fmt(
+    data = data,
+    columns = {{ columns }},
+    rows = {{ rows }},
+    fns = num_fmt_factory_multi(
+      pattern = pattern,
+      format_fn = function(x, context) {
+
+        # Round all values of x to 3 digits with the R-H-U method of
+        # rounding (for reproducibility purposes)
+        x <- round_gt(x, 3)
+
+        # Determine which of `x` are finite values
+        x_is_a_number <- is.finite(x)
+
+        # Divide the `x` values in 'big' and 'small' components; delay the
+        # formatting of `big_x` until it is appropriately rounded on the
+        # basis of the fractions obtained at the desired accuracy
+        big_x <- trunc(x)
+        small_x <- abs(x - big_x)
+
+        # Get the correct minus mark based on the output context
+        minus_mark <- context_minus_mark(context = context)
+
+        # Format the 'small' portion of the numeric values
+        # to character-based numbers with exactly 3 decimal places
+        small_x[x_is_a_number] <-
+          format_num_to_str(
+            small_x[x_is_a_number],
+            context = context, decimals = 3, n_sigfig = NULL,
+            sep_mark = ",", dec_mark = ".",
+            drop_trailing_zeros = FALSE,
+            drop_trailing_dec_mark = TRUE,
+            format = "f"
+          )
+
+        # Map the `accuracy` keyword to a column index in the
+        # internal `fractions` dataset
+        fractions_col_idx <-
+          which(c("low", "med", "high", "/2", "/4", "/8", "/16", "/100") %in% accuracy) + 1
+
+        # Generate an vector of empty strings that will eventually contain
+        # all of the fractional parts of the finalized numbers
+        fraction_x <- rep("", length(x))
+
+        # For every `small_x` value that corresponds to a number
+        # (i.e., not Inf), get the fractional part from the `fractions`
+        # lookup table
+        fraction_x[x_is_a_number] <-
+          vapply(
+            small_x[x_is_a_number],
+            FUN.VALUE = character(1),
+            USE.NAMES = FALSE,
+            FUN = function(x) {
+              fractions[fractions[["x"]] == x, ][[fractions_col_idx]]
+            }
+          )
```

Comment:
This vapply can be replaced by just a single `[`

Reply:
This is now much simplified.